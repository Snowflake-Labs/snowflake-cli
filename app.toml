snowsql_config_path = "/Users/zblackwood/.snowsql/config"
snowsql_connection_name = "snowhouse"

[dev]
database = "STREAMLIT"
schema = "SIS_APPS"
role = "STREAMLIT_MODELING_RL"
warehouse = "SNOWHOUSE"
