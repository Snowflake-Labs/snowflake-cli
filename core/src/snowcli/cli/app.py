from __future__ import annotations

import importlib
import logging
from pathlib import Path

import click
import typer

from snowcli import __about__
from snowcli.cli import loggers
from snowcli.cli.common.snow_cli_global_context import (
    SnowCliGlobalContext,
    snow_cli_global_context_manager,
    global_context_copy,
)
from snowcli.cli.main.snow_cli_main_typer import SnowCliMainTyper
from snowcli.config import config_init, cli_config
from snowcli.docs.generator import generate_docs
from snowcli.output.formats import OutputFormat
from snowcli.output.printing import print_data
from snowcli.plugin.load_plugins import PluginLoadingMode
from snowcli.plugin.plugin_registration import load_and_register_plugins_in_typer
from snowcli.pycharm_remote_debug import setup_pycharm_remote_debugger_if_provided

app: SnowCliMainTyper = SnowCliMainTyper()
log = logging.getLogger(__name__)


def _docs_callback(value: bool):
    if value:
        ctx = click.get_current_context()
        generate_docs(Path("gen_docs"), ctx.command)
        raise typer.Exit()


def _version_callback(value: bool):
    if value:
        typer.echo(f"SnowCLI Version: {__about__.VERSION}")
        raise typer.Exit()


def _info_callback(value: bool):
    if value:
        print_data(
            [
                {"key": "version", "value": __about__.VERSION},
                {"key": "default_config_file_path", "value": cli_config.file_path},
            ]
        )
        raise typer.Exit()


def _debug_callback(debug: bool):
    def modifications(context: SnowCliGlobalContext) -> SnowCliGlobalContext:
        context.debug = debug
        return context

    snow_cli_global_context_manager.update_global_context(modifications)


def _load_all_plugins_mode_callback(load_all_plugins_mode: bool):
    def modifications(context: SnowCliGlobalContext) -> SnowCliGlobalContext:
        context.plugin_loading_mode = (
            PluginLoadingMode.ALL_INSTALLED_PLUGINS
            if load_all_plugins_mode
            else PluginLoadingMode.ONLY_ENABLED_PLUGINS
        )
        return context

    snow_cli_global_context_manager.update_global_context(modifications)


def _load_and_register_plugins(enable_plugins: bool) -> None:
    if enable_plugins:
        plugin_loading_mode = global_context_copy().plugin_loading_mode
        load_and_register_plugins_in_typer(plugin_loading_mode)


@app.callback()
def default(
    debug: bool = typer.Option(
        None,
        "--debug",
        help="Print logs from level debug and higher, logs contains additional information",
        is_eager=True,
        callback=_debug_callback,
    ),
    load_all_plugins: bool = typer.Option(
        False,
        "--load-all-plugins",
        help="Load all installed plugins (even these which are explicitly disabled)",
        is_eager=True,
        callback=_load_all_plugins_mode_callback,
    ),
    configuration_file: Path = typer.Option(
        None,
        "--config-file",
        help="Specifies snowcli configuration file that should be used",
        exists=True,
        dir_okay=False,
        is_eager=True,
        callback=config_init,
    ),
    # This hidden option is required to register plugins before other options and commands evaluation.
    # It has to be placed under "safe", "load_all_plugins"
    # and "configuration_file" options to have access to their values.
    enable_plugins: bool = typer.Option(
        True,
        "--enable-plugins",
        help="Enable plugins",
        hidden=True,
        is_eager=True,
        callback=_load_and_register_plugins,
    ),
    version: bool = typer.Option(
        None,
        "--version",
        help="Shows version of the snowcli",
        callback=_version_callback,
        is_eager=True,
    ),
    docs: bool = typer.Option(
        None,
        "--docs",
        hidden=True,
        help="Generates Snowflake CLI documentation",
        callback=_docs_callback,
        is_eager=True,
    ),
    info: bool = typer.Option(
        None,
        "--info",
        help="Shows information about the snowcli",
        callback=_info_callback,
    ),
    output_format: OutputFormat = typer.Option(
        OutputFormat.TABLE.value,
        "--format",
        help="Specifies output format",
        case_sensitive=False,
        is_eager=True,
    ),
    verbose: bool = typer.Option(
        None,
        "--verbose",
        "-v",
        help="Print logs from level info and higher",
    ),
    pycharm_debug_library_path: str = typer.Option(
        None,
        "--pycharm-debug-library-path",
        hidden=True,
    ),
    pycharm_debug_server_host: str = typer.Option(
        "localhost",
        "--pycharm-debug-server-host",
        hidden=True,
    ),
    pycharm_debug_server_port: int = typer.Option(
        12345,
        "--pycharm-debug-server-port",
        hidden=True,
    ),
) -> None:
    """
    SnowCLI - A CLI for Snowflake
    """
    setup_pycharm_remote_debugger_if_provided(
        pycharm_debug_library_path=pycharm_debug_library_path,
        pycharm_debug_server_host=pycharm_debug_server_host,
        pycharm_debug_server_port=pycharm_debug_server_port,
    )
    loggers.create_loggers(verbose, debug)


def _add_typer_from_path(path: str):
    sub_module = importlib.import_module(path)
    app.add_typer(getattr(sub_module, "app"))


def _register_internal_cli_typers() -> None:
    known_sub_commands = [
        "snowcli.cli.snowpark",
        "snowcli.cli.render",
        "snowcli.cli.streamlit",
        "snowcli.cli.warehouse",
        "snowcli.cli.stage.commands",
        "snowcli.cli.plugin_management",
    ]
    for cmd in known_sub_commands:
        _add_typer_from_path(cmd)

    from snowcli.cli import sql

    app.command("sql")(sql.execute_sql)


_register_internal_cli_typers()
