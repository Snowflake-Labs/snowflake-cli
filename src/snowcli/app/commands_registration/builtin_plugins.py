from snowcli.plugins.connection import plugin_spec as connection_plugin_spec
from snowcli.plugins.containers import plugin_spec as containers_plugin_spec
from snowcli.plugins.nativeapp import plugin_spec as nativeapp_plugin_spec
from snowcli.plugins.object import plugin_spec as object_plugin_spec
from snowcli.plugins.registry import plugin_spec as registry_plugins_spec
from snowcli.plugins.render import plugin_spec as render_plugin_spec
from snowcli.plugins.snowpark import plugin_spec as snowpark_plugin_spec
from snowcli.plugins.sql import plugin_spec as sql_plugin_spec
from snowcli.plugins.streamlit import plugin_spec as streamlit_plugin_spec

# plugin name to plugin spec
builtin_plugin_name_to_plugin_spec = {
    "connection": connection_plugin_spec,
    "containers": containers_plugin_spec,
    "nativeapp": nativeapp_plugin_spec,
    "object": object_plugin_spec,
    "registry": registry_plugins_spec,
    "render": render_plugin_spec,
    "snowpark": snowpark_plugin_spec,
    "sql": sql_plugin_spec,
    "streamlit": streamlit_plugin_spec,
}
