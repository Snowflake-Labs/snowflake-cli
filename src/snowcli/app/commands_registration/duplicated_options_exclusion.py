from __future__ import annotations

from dataclasses import dataclass
from typing import List, Union

import click
from click import Command, Parameter
from typer.core import TyperGroup, TyperCommand


@dataclass
class SecondaryOptionNameExclusion:
    command_path: str
    primary_option_name: str
    excluded_secondary_option_name: str


_secondary_option_names_duplicates = [
    SecondaryOptionNameExclusion("snowpark procedure describe", "--password", "-p"),
    SecondaryOptionNameExclusion("snowpark procedure drop", "--password", "-p"),
    SecondaryOptionNameExclusion("snowpark procedure execute", "--password", "-p"),
    SecondaryOptionNameExclusion("snowpark services create", "--password", "-p"),
]


def exclude_duplicated_secondary_option_names():
    _exclude_secondary_option_names(_secondary_option_names_duplicates)


def _exclude_secondary_option_names(exclusions: List[SecondaryOptionNameExclusion]):
    main_command_group = _get_main_typer_command_group_from_click_context()
    for exclusion in exclusions:
        command = _find_command(
            command_group=main_command_group,
            remaining_path=exclusion.command_path.split(" "),
            full_path=exclusion.command_path,
        )
        param = _find_param(
            command=command,
            primary_option_name=exclusion.primary_option_name,
            command_path=exclusion.command_path,
        )
        _remove_option_name_from_param(exclusion.excluded_secondary_option_name, param)


def _get_main_typer_command_group_from_click_context() -> TyperGroup:
    main_typer_command_group = click.get_current_context().command
    if isinstance(main_typer_command_group, TyperGroup):
        return main_typer_command_group
    else:
        raise RuntimeError(
            "Invalid main top-level command type. It should be a TyperGroup but it is not."
        )


def _find_command(
    command_group: TyperGroup,
    remaining_path: List[str],
    full_path: str,
) -> Union[TyperCommand, TyperGroup]:
    if remaining_path:
        new_remaining_path = remaining_path[1:]
        found_command = command_group.commands.get(remaining_path[0])
        if isinstance(found_command, TyperGroup):
            return _find_command(found_command, new_remaining_path, full_path)
        elif isinstance(found_command, TyperCommand) and not new_remaining_path:
            return found_command
        else:
            raise RuntimeError(f"Cannot find command [{full_path}].")
    else:
        return command_group


def _find_param(
    command: Command, primary_option_name: str, command_path: str
) -> Parameter:
    for param in command.params:
        if primary_option_name in param.opts:
            return param
    raise RuntimeError(
        f"Cannot find param [{primary_option_name}] for command [{command_path}]."
    )


def _remove_option_name_from_param(excluded_name: str, param: Parameter):
    if excluded_name in param.opts:
        param.opts.remove(excluded_name)
    if excluded_name in param.secondary_opts:
        param.secondary_opts.remove(excluded_name)
