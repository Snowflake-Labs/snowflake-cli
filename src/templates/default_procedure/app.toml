# The values in this file will be used if flags aren't given
# in CLI commands. Can create multiple --environments (default 'default')

[connections]
[connections.default]
input_parameters = "()"
return_type = "string"
name = "helloFunction"
handler = "app.hello"
file = "app.zip"

[connections.dev]
database = "<DATABASE_NAME>"
schema = "<SCHEMA_NAME>"
warehouse = "<WAREHOUSE_NAME>"
role = "<ROLE_NAME>"
