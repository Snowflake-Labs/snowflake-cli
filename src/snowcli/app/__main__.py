from __future__ import annotations

import sys

from snowcli.app.cli_app import app


def main(*args):
    if not sys.warnoptions:
        import warnings

        warnings.filterwarnings("ignore", module="snowflake\.connector")
    app(*args)


if __name__ == "__main__":
    main()

if getattr(sys, "frozen", False):
    main(sys.argv[1:])
