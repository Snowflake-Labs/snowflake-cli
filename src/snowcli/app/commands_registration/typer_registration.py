import logging
from typing import List

import click
from typer.core import TyperGroup

from snowcli.app.commands_registration import LoadedCommandPlugin
from snowcli.api.plugin.command import CommandSpec
from snowcli.cli.exception_logging import exception_logging

log = logging.getLogger(__name__)
log_exception = exception_logging(log)


class TyperCommandsRegistration:
    def __init__(self, plugins: List[LoadedCommandPlugin]):
        self._plugins = plugins
        self._main_typer_command_group = (
            self._get_main_typer_command_group_from_click_context()
        )

    def register_commands(self):
        for plugin in self._plugins:
            try:
                self._add_plugin_to_typer(plugin.command_spec)
            except Exception as ex:
                log_exception(
                    f"Cannot register plugin [{plugin.plugin_name}]: {ex.__str__()}", ex
                )

    @staticmethod
    def _get_main_typer_command_group_from_click_context() -> TyperGroup:
        main_typer_command_group = click.get_current_context().command
        if isinstance(main_typer_command_group, TyperGroup):
            return main_typer_command_group
        else:
            raise RuntimeError(
                "Invalid main top-level command type. It should be a TyperGroup but it is not."
            )

    def _add_plugin_to_typer(
        self,
        command_spec: CommandSpec,
    ) -> None:
        parent_group = self._find_typer_group_at_path(
            current_level_group=self._main_typer_command_group,
            remaining_parent_path_segments=command_spec.parent_command_path.path_segments,
            command_spec=command_spec,
        )
        command = command_spec.command
        if command.name in parent_group.commands:
            raise RuntimeError(
                f"Cannot add command [{command_spec.full_command_path}] because it already exists."
            )
        parent_group.add_command(command)

    def _find_typer_group_at_path(
        self,
        current_level_group: TyperGroup,
        remaining_parent_path_segments: List[str],
        command_spec: CommandSpec,
    ) -> TyperGroup:
        if remaining_parent_path_segments:
            expected_name = remaining_parent_path_segments[0]
            matching_subgroups = [
                subgroup
                for subgroup in current_level_group.commands.values()
                if isinstance(subgroup, TyperGroup) and subgroup.name == expected_name
            ]
            if matching_subgroups:
                return self._find_typer_group_at_path(
                    current_level_group=matching_subgroups[0],
                    remaining_parent_path_segments=remaining_parent_path_segments[1:],
                    command_spec=command_spec,
                )
            else:
                raise RuntimeError(
                    f"Invalid command path [{command_spec.full_command_path}]. "
                    f"Command group [{expected_name}] does not exist."
                )
        else:
            return current_level_group


def register_commands_from_plugins(plugins: List[LoadedCommandPlugin]) -> None:
    return TyperCommandsRegistration(plugins).register_commands()
