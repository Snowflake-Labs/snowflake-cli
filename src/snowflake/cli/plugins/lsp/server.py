# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Callable

from pygls.server import LanguageServer

from snowflake.cli.__about__ import VERSION

def lsp_plugin(name: str, capabilities: dict) -> Callable:
    """
    Defines the root of an LSP plugin.
    The intention is for each Snowflake CLI plugin to have at most 1 LSP plugin (namespace).
    """
    def _decorator(func: Callable) -> Callable:
        setattr(
            func,
            "_lsp_context",
            {
                "lsp_plugin_name": name,
                "lsp_plugin_capabilities": capabilities,
            },
        )
        return func

    return _decorator


def start_lsp_server():
    """
    Find and initialize all LSP plugins, then start the LSP server.    
    """

    # N.B. LSP server version is aligned with Snowflake CLI version
    server = LanguageServer(name="lsp_controller", version=VERSION)

    for plugin_func in iter_lsp_plugins():
        plugin_func.lsp_spec(server)

    server.start_io()
    return server


def iter_lsp_plugins():
    """
    Dynamically discover all plugin functions that implement the LSP spec
    (i.e. all functions decorated with @lsp_plugin).
    """

    # Use a dynamic import to avoid semi-circular imports if at top level
    from snowflake.cli.app.commands_registration.command_plugins_loader import (
        load_only_builtin_command_plugins,
    )

    plugins = load_only_builtin_command_plugins()
    for plugin_func in plugins:
        if getattr(plugin_func, "lsp_spec", None) is not None:
            yield plugin_func


