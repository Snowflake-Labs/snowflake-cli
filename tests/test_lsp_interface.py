# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from unittest import mock

import pytest
from snowflake.cli.app.commands_registration import LoadedBuiltInCommandPlugin
from snowflake.cli.plugins.lsp.server import iter_lsp_plugins
from typing import get_type_hints, Any, Iterator, Dict, Callable

from pygls.server import LanguageServer

WORKSPACE_COMMAND = 'snowflake.cli.plugins.lsp.interface.workspace_command'


@pytest.fixture
def mock_workspace_command() -> Iterator[Dict[str, Callable]]:
    found_commands: Dict[str, Callable] = dict() 

    def workspace_command_impl(_server: Any, command_name: str):
        print("bleh", command_name)
        def _wrapper(fn: Callable):
            found_commands[command_name] = fn
        return _wrapper

    with mock.patch(WORKSPACE_COMMAND) as mocker:
        mocker.side_effect = workspace_command_impl
        mocker.found_commands = found_commands
        yield mocker


@pytest.mark.parametrize(
    "plugin",
    iter_lsp_plugins(),
    ids=lambda x: x.plugin_name,
)
def test_command_signatures(snapshot, mock_workspace_command, plugin: LoadedBuiltInCommandPlugin):
    """
    Compare command signatures to snapshots.
    """
    lsp_context = plugin.lsp_spec._lsp_context
    assert isinstance(lsp_context, dict)

    # first, call the lsp spec function to populate commands
    # we use mock_workspace_command to enumerate without doing any real work
    server = mock.Mock(spec=LanguageServer)
    plugin.lsp_spec(server)

    # make sure there are no missing / extra command implementations
    capabilities: dict = lsp_context["lsp_plugin_capabilities"]
    found_commands: dict = mock_workspace_command.found_commands
    assert sorted(found_commands.keys()) == sorted(capabilities.keys())

    # for each found workspace command, snapshot its type signature    
    signatures = { k: get_type_hints(v) for k, v in found_commands.items() }
    assert signatures == snapshot

    # TODO: this does not account for changes to the types themselves
    # we should recursively snapshot any typed dicts / dataclasses /
    # whatever will automatically be marshalled to / from JSON
