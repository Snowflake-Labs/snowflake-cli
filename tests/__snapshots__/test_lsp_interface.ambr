# serializer version: 1
# name: test_command_signatures[nativeapp]
  dict({
    'openApplication': dict({
      'connection': <class 'snowflake.connector.connection.SnowflakeConnection'>,
      'project_path': <class 'str'>,
      'return': <class 'snowflake.cli.api.output.types.MessageResult'>,
    }),
  })
# ---
